src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/CancellationTokenSourceFactory.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Roslyn.Utilities
+{
+    internal static class CancellationTokenSourceFactory
+    {
+        private static readonly ConditionalWeakTable<CancellationTokenSource, Tuple<string, int>> s_tokenCreators = new();
+
+        private static CancellationTokenSource RecordLocation(CancellationTokenSource source, string? sourcePath, int sourceLine)
+        {
+            Contract.ThrowIfNull(sourcePath);
+            s_tokenCreators.Add(source, Tuple.Create(sourcePath, sourceLine));
+            return source;
+        }
+
+        public static CancellationTokenSource Create([CallerFilePath] string? sourcePath = null, [CallerLineNumber] int sourceLine = 0)
+            => RecordLocation(new CancellationTokenSource(), sourcePath, sourceLine);
+
+#pragma warning disable CA1068 // CancellationToken parameters must come last
```

Comment:
📝 This was recently fixed with https://github.com/dotnet/roslyn-analyzers/pull/4228